train.py

```diff
@@ -2,84 +2,135 @@
 
-                metric_accum.gen_accum(loss.item())
+        for epoch in range(epochs_num):
+            for batch_num, X_batch_real in enumerate(tqdm(train_loader, desc=f'epoch {epoch}', position=0, leave=True)):
+                self.generator.train()
+                batch_size = X_batch_real.size(0)
+
+                if (iterations_total + 1) % parameters.save_every == 0:
+                    save_model(save_dir, self.generator, self.discriminator, optimizer_g, optimizer_d, iterations_total)
+                if iterations_total > parameters.iterations:
+                    break
+
+                iterations_total += 1
+
+                for _ in range(parameters.training_ratio):
+                    d_loss = self.discriminator_loss(parameters, X_batch_real)
+
+                    optimizer_d.zero_grad()
+                    d_loss.backward()
+                    optimizer_d.step()
+
+                    if scheduler_d is not None:
+                        scheduler_d.step()
+
+                g_loss = self.generator_loss((batch_size, parameters.gan_noise_size))
 
                 optimizer_g.zero_grad()
-                loss.backward()
+                g_loss.backward()
                 optimizer_g.step()
 
                 if scheduler_g is not None:
                     scheduler_g.step()
 
-            X_batch_real = X_batch_real.to(device)
+                if experiment is not None and iterations_total % parameters.log_every == 0:
+                    assert test_dataset is not None
+                    experiment.log_metrics({'g_loss': g_loss.detach().cpu().numpy(),
+                                            'd_loss': d_loss.detach().cpu().numpy()})
+                    eval_batch_num = int((parameters.gan_test_ratio * len(test_dataset)) / parameters.eval_batch_size)
+                    evaluate_model(self.generator, experiment,
+                                   test_dataset, parameters.eval_batch_size,
+                                   eval_batch_num, parameters,
+                                   self.device, scaler, iterations_total)
+
+        return iterations_total
+
+    def discriminator_loss(self, parameters, X_batch_real):
+        X_batch_real = X_batch_real.to(self.device)
+
+        X_noise = torch.randn((X_batch_real.size(0), parameters.gan_noise_size)).to(self.device)
+        G_output = self.generator(X_noise)
+
+        D_fake = self.discriminator(G_output.float())
+        D_real = self.discriminator(X_batch_real.float())
+
+        D_loss = -torch.mean(torch.cat((torch.log(D_real + 1e-8),
+                                        torch.log(1 - D_fake + 1e-8))))
+
+        return D_loss
+
+    def generator_loss(self, noise_batch_shape):
+        X_noise = torch.randn(noise_batch_shape).to(self.device)
+        G_output = self.generator(X_noise)
+        D_output = self.discriminator(G_output.float())
+
+        G_loss = -torch.mean(torch.log(D_output + 1e-8))
+
+        return G_loss
```

Comment:
For vanilla GAN we have implemented functions or classes in torch for BCELoss calculations, that optimized and works well, what reason for implementing it manually?.

Reply:
To provide the same interface for all classes.

Reply:
I meant something like
```
from torch.nn.functional import binary_cross_entropy

def generator_loss(self, noise_batch_shape):
    X_noise = torch.randn(noise_batch_shape, device=device)
    G_output = self.generator(X_noise)
    D_output = self.discriminator(G_output.float())
    target = torch.ones_like(D_output, device=device)

    return binary_cross_entropy(D_output, target)
```
this more obvious when you reading code
But I agree this is not important in this case actually, so do not change it if you think this is better now.